src/main/clojure/jarl/builtins/jwt.clj

```diff
@@ -0,0 +1,308 @@
+            [jarl.encoding.json :as json]
+            [jarl.encoding.hex :as hex]
+            [jarl.exceptions :as errors])
+  (:import (org.jose4j.jwk JsonWebKeySet JsonWebKey$Factory JsonWebKey RsaJsonWebKey EllipticCurveJsonWebKey)
+           (org.jose4j.keys.resolvers JwksVerificationKeyResolver VerificationKeyResolver X509VerificationKeyResolver)
+           (org.jose4j.keys X509Util HmacKey)
+           (java.util List Collections Map)
+           (org.jose4j.jws JsonWebSignature)
+           (java.security KeyFactory)
+           (java.security.spec X509EncodedKeySpec)
+           (java.security.interfaces RSAPublicKey ECPublicKey)
+           (clojure.lang ExceptionInfo)
+           (org.jose4j.lang UnresolvableKeyException JoseException)))
+
+(def supported-algs #{"HS256" "HS384" "HS512" "ES256" "ES384" "ES512" "RS256" "RS384" "RS512" "PS256" "PS384" "PS512"})
+
+(defn alg-str->alg [s]
+  (when (some? s)
+    (if (keyword? s) s (keyword (str/lower-case s)))))
+
+(defn- split-jwt [jwt]
+  (let [parts (str/split jwt #"\.") num-parts (count parts)]
+    (if (= 3 num-parts)
+      parts
+      (if (= 1 num-parts)
+        (throw (errors/builtin-ex "encoded JWT had no period separators"))
+        (throw (errors/builtin-ex "encoded JWT must have 3 sections, found %s" num-parts))))))
+
+(defn- strip-pem-wrapper [^String s]
+  (if (str/starts-with? s "---")
+    (str/join "\n" (-> (str/split-lines s) rest butlast))
+    s))
+
+(defn- base64-decode-bytes ^bytes [^String s part]
+  (let [strip-ex-type (fn [message] (subs message (+ 2 (str/index-of message ":"))))]
+    (try
+      (base64/url-decode-bytes s)
+      (catch ExceptionInfo e
+        (if (errors/builtin-ex? e)
+          (throw (errors/builtin-ex "JWT %s had invalid encoding: %s" part (strip-ex-type (ex-message e))))
+          (throw e))))))
+
+(defn jwk-key-resolver ^VerificationKeyResolver [^String s]
+  (reify VerificationKeyResolver
+    (resolveKey [_ _ _]
+      (try
+        (.getKey (JsonWebKey$Factory/newJwk s))
+        (catch Throwable _
+          (throw (errors/builtin-ex "failed to parse a JWK key (set)")))))))
+
+(defn strict-jwks-key-resolver ^VerificationKeyResolver [^String s]
+  (-> (JsonWebKeySet. s)
+      (.getJsonWebKeys)
+      (JwksVerificationKeyResolver.)
+      (doto (.setDisambiguateWithVerifySignature true))))
+
+(defn valid? [^JsonWebSignature jws ^JsonWebKey key]
+  (and (= (.getAlgorithmHeaderValue jws) (.getAlgorithm key))
+       (.verifySignature (doto jws (.setKey (.getKey key))))))
+
+(defn- try-strict-resolve [^String s jws nesting-context]
+  (try
+    (.resolveKey (strict-jwks-key-resolver s) jws nesting-context)
+    (catch UnresolvableKeyException _ false)))
+
+(defn jwks-key-resolver [^String s]
+  (reify VerificationKeyResolver
+    (resolveKey [_ jws nesting-context]
+      (if-let [key (try-strict-resolve s jws nesting-context)]
+        (.getKey ^JsonWebKey key)
+        (.getKey ^JsonWebKey (first (filter #(valid? jws %)
```

Comment:
Doesn't OPA also consider the `kid`?

Reply:
The `try-strict-resolve` function does this, so if the key is found that way we won't filter over the list of keys.